### PR TITLE
New @JsonQualifier annotation.

### PR DESCRIPTION
**Pull request by swankjesse**
Works like JSR-330's @Qualifier annotation. You may have
multiple qualifiers, or none.

### CODE REVIEW

File: moshi/src/main/java/com/squareup/moshi/Util.java [hunk 1]
```diff
@@ -18,27 +18,41 @@
...
-      return null;
-    }
-    @Override public Annotation[] getAnnotations() {
-      return EMPTY_ANNOTATIONS_ARRAY;
-    }
-    @Override public Annotation[] getDeclaredAnnotations() {
-      return EMPTY_ANNOTATIONS_ARRAY;
-    }
-  };
+  public static final Set<Annotation> NO_ANNOTATIONS = Collections.emptySet();
```
Comment (JakeWharton) [inline moshi/src/main/java/com/squareup/moshi/Util.java]: Maybe just inline? You're using `Collections.<Annotation>emptySet()` elsewhere.

File: moshi/src/test/java/com/squareup/moshi/CircularAdaptersTest.java [hunk 2]
```diff
@@ -97,19 +100,20 @@ Node minusPrefix(String prefix) {
...
    */
   static class PrefixingNodeFactory implements JsonAdapter.Factory {
-    @Override public JsonAdapter<?> create(Type type, AnnotatedElement annotations, Moshi moshi) {
+    @Override public JsonAdapter<?> create(
+        Type type, Set<? extends Annotation> annotations, Moshi moshi) {
       if (type != Node.class) return null;
 
       final String prefix;
-      if (annotations.isAnnotationPresent(Left.class)) {
+      if (Util.isAnnotationPresent(annotations, Left.class)) {
```
Comment (JakeWharton) [inline moshi/src/test/java/com/squareup/moshi/CircularAdaptersTest.java]: Something to think about: this convenience method won't be available to those writing custom adapter factories.

File: moshi/src/test/java/com/squareup/moshi/ClassJsonAdapterTest.java [hunk 3]
```diff
@@ -25,7 +25,6 @@
 import org.junit.Test;
 
 import static com.squareup.moshi.TestUtil.newReader;
-import static com.squareup.moshi.Util.NO_ANNOTATIONS;
```
Comment (JakeWharton) [inline moshi/src/test/java/com/squareup/moshi/ClassJsonAdapterTest.java]: Revert this whole file's changes (or just static import so they match). History noise.

File: moshi/src/test/java/com/squareup/moshi/JsonQualifiersTest.java [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+          + "does not have @JsonQualifier");
+    }
+  }
+
+  @Test public void annotationsConflict() throws Exception {
+    try {
+      new Moshi.Builder().add(new AnnotationsConflictJsonAdapter());
+      fail();
+    } catch (IllegalArgumentException expected) {
+      assertThat(expected.getMessage()).contains("Conflicting @ToJson methods");
```
Comment (JakeWharton) [inline moshi/src/test/java/com/squareup/moshi/JsonQualifiersTest.java]: I think there's also `hasMessageStartingWith`. AssertJ is crazy...

**Comment by JakeWharton** [comment]
LGTM! Awesome stuff.

File: moshi/src/main/java/com/squareup/moshi/Util.java (hunk 1 above)
Comment (swankjesse) [inline moshi/src/main/java/com/squareup/moshi/Util.java]: I did the opposite, changing callers to `Util.NO_ANNOTATIONS`. It's got ~12 uses. I'm open to flip flop again on this.

File: moshi/src/test/java/com/squareup/moshi/CircularAdaptersTest.java (hunk 2 above)
Comment (swankjesse) [inline moshi/src/test/java/com/squareup/moshi/CircularAdaptersTest.java]: Yup. My hope is that most developers will be using adapter methods anyway.

File: moshi/src/test/java/com/squareup/moshi/ClassJsonAdapterTest.java (hunk 3 above)
Comment (swankjesse) [inline moshi/src/test/java/com/squareup/moshi/ClassJsonAdapterTest.java]: Good call. I originally replaced it with a bunch of `Set`s, and when I didn't like that I unintentionally changed the imports.

File: moshi/src/test/java/com/squareup/moshi/JsonQualifiersTest.java (hunk 4 above)
Comment (swankjesse) [inline moshi/src/test/java/com/squareup/moshi/JsonQualifiersTest.java]: Found it. Fixed.

File: moshi/src/main/java/com/squareup/moshi/Util.java (hunk 1 above)
Comment (JakeWharton) [inline moshi/src/main/java/com/squareup/moshi/Util.java]: Consistency was what I was advocating so :+1:

**Pull request merged**